# Visualising Results
## Overview
In this `.Rmd` we are visualising many of the outputs from the SNP- and SV-based population analyses for fairy tern populations in Western Australian and Aotearoa New Zealand.   
```{r setup, include=FALSE}
setwd("C:/Users/Jana/Desktop/tara_iti")

library(tidyr)
library(ggplot2)
library(dplyr)
library(mclust)
#install.packages("devtools")
#devtools::install_github("G-Thomson/Manu")
library(Manu)

pal <- get_pal("Kotare")
get_pal("Kotare")
print_pal(pal)
```
## Heterozygosity
Comparisons of individual heterozygosity as estimated by ANGSD.  
```{r heterozygosity, include=FALSE}
indiv_het <- read.csv("angsd/individual_het.tsv", sep="\t", header = TRUE)

het_fig <- filter(indiv_het, Tool == 'samtools') %>%
    ggplot(aes(x=Population, y = Heterozygosity, fill = Population)) +
        geom_violin() +
        theme_light() +
        labs(x = "Population", y = "Observed Heterozygosity", title = "Individual Heterozygosity") +
        scale_fill_manual(values = c("AU" = "#E5BF86","NZ" = "#287DAB")) 
ggsave('plots/SNP_samtools_heterozygosity2.png', het_fig, device = "png", dpi = 600)
```

### Runs of Homozygosity
Create a one-column file (e.g., scat_rohs.txt) with ROH sizes across all samples. 
```{r ROH clusters, include=FALSE}
nzroh <- read.csv("nz_roh.txt", header = FALSE)
nzroh_clust <- Mclust(nzroh)
nzroh_bic <- mclustBIC(nzroh)

auroh <- read.csv("au_roh.txt", header = FALSE)
auroh_clust <- Mclust(auroh)
auroh_bic <- mclustBIC(auroh)

summary(nzroh_clust, parameter=TRUE)
summary(nzroh_bic, parameter=TRUE)
summary(auroh_clust, parameter=TRUE)
summary(auroh_bic, parameter=TRUE)

x <- seq(0, 8000000, length = 1000)

nzstern1 <- dnorm(x, mean=nzroh_clust$parameters$mean[1], sd=sqrt(nzroh_clust$parameters$variance$sigmasq[1]))
nzstern2 <- dnorm(x, mean=nzroh_clust$parameters$mean[2], sd=sqrt(nzroh_clust$parameters$variance$sigmasq[2]))
nzstern3 <- dnorm(x, mean=nzroh_clust$parameters$mean[3], sd=sqrt(nzroh_clust$parameters$variance$sigmasq[3]))

austern1 <- dnorm(x, mean=auroh_clust$parameters$mean[1], sd=sqrt(auroh_clust$parameters$variance$sigmasq[1]))
austern2 <- dnorm(x, mean=auroh_clust$parameters$mean[2], sd=sqrt(auroh_clust$parameters$variance$sigmasq[2]))
austern3 <- dnorm(x, mean=auroh_clust$parameters$mean[3], sd=sqrt(auroh_clust$parameters$variance$sigmasq[3]))
```
Here we plot the clustering of ROHs in tara iti.  

```{r nz clust plots, include=FALSE}
png("plots/Tara_iti_ROH_BIC.png")
plot(nzroh_bic)
dev.off()

png("plots/Tara_iti_ROH_frequency.png")
hist(nzroh_clust$data, breaks=132, main="Distribution of ROH for tara iti", xlab="Number of base pairs")
lines(x, (nzstern1*1800000000), col="red")
lines(x, (nzstern2*1800000000), col="blue")
lines(x, (nzstern3*1800000000), col="green")
legend("topright", title="Gaussian groups", legend=c("Short","Medium","Long"), lty=c(1,1,1), col=c("red", "blue", "green"))
dev.off()

png("plots/Tara_iti_ROH_sizes.png")
hist(nzroh_clust$data, breaks = 132, main = "Distribution of ROH for S. nereis davisae", xlab = "Number of base pairs")
abline(v = 200000, lty = 5, col = "red")
abline(v = 700000, lty = 4, col = "blue")
legend("topright", title="ROH boundries", legend=c("Short-Medium break","Medium-Long break"), lty=c(5,4), col=c("red", "blue"))
dev.off()

png(filename="plots/Tara_iti_ROH_clusters.png")
plot(x, nzstern1, col = "red", type = "l", main = "ROH length distributions for S. nereis davisae",
     xlab = "Number of base pairs", ylab = "", yaxt = "n")
lines(x, (nzstern2), col = "blue")
lines(x, (nzstern3), col = "green")
legend("topright", title = "Gaussian groups", legend = c("Short", "Medium", "Long"), lty = c(1,1,1), col = c("red", "blue", "green"))
dev.off()
```

```{r au clust plots, include=FALSE}
png(filename = "plots/Australia_ROH_BIC.png")
plot(auroh_bic)
dev.off()

png(filename = "plots/Australia_ROH_frequency.png")
hist(auroh_clust$data, breaks=132, main="Distribution of ROH for Australian fairy tern", xlab="Number of base pairs")
lines(x, (austern1*1800000000), col="red")
lines(x, (austern2*1800000000), col="blue")
lines(x, (austern3*1800000000), col="green")
legend("topright", title="Gaussian groups", legend=c("Short","Medium","Long"), lty=c(1,1,1), col=c("red", "blue", "green"))
dev.off()

png(filename = "plots/Australia_ROH_sizes.png")
hist(auroh_clust$data, breaks = 132, main = "Distribution of ROH for S. nereis nereis", xlab = "Number of base pairs")
abline(v = 200000, lty = 5, col = "red")
abline(v = 700000, lty = 4, col = "blue")
legend("topright", title="ROH boundries", legend=c("Short-Medium break","Medium-Long break"), lty=c(5,4), col=c("red", "blue"))
dev.off()

png(filename="plots/Australia_ROH_clusters.png")
plot(x, austern1, col = "red", type = "l", main = "ROH length distributions for S. nereis davisae",
     xlab = "Number of base pairs", ylab = "", yaxt = "n")
lines(x, (austern2), col = "blue")
lines(x, (austern3), col = "green")
legend("topright", title = "Gaussian groups", legend = c("Short", "Medium", "Long"), lty = c(1,1,1), col = c("red", "blue", "green"))
dev.off()
```
Temporary ROH Stats
```{r ROH plots and summary, include=FALSE}
ROH <- read.csv('ROH_summary.txt', sep='\t', header = TRUE)
head(ROH)
filter(ROH, Population == 'AU') %>% summary()
filter(ROH, Population == 'NZ') %>% summary()

ROH_size_plot <- ggplot(ROH, aes(x = Population, y = Mean_ROH_size, fill = Population)) +
    geom_violin() +
    theme_light() +
    labs(x = "Population", y = "Mean ROH Size") +
    scale_fill_manual(values = c("AU" = "#E5BF86","NZ" = "#287DAB"))

ggsave( 'plots/average_ROH_length.png', ROH_size_plot, device = "png")
```

# Site Frequency Spectrum (SFS)
## Individual SFS
First we plotted the SFS for Australian fairy tern and tara iti for comparison. Here we import the results of the Australian fairy tern (`AU.sfs`) and tara iti (`TI.sfs`) site frequency spectrum output from `realSFS` and filtered to include variable alleles.  
```{r indiv SFS, include=FALSE}
nnorm <- function(x) x/sum(x)

sfs <- rbind(
    AU=scan("AU.sfs")[-1],
    NZ=scan("TI.sfs")[-1])
colnames(sfs) <- 1:19
```
We then plot the count density.
```{r indiv SFS density, include=FALSE}
barplot(sfs, beside=TRUE, legend=("AU", "NZ"), names=1:19, main = "realSFS non-ancestral sites", args.legend = list(x="topleft"))
```
Then we plotted the polymorphic sites.
```{r indiv SFS plot, include=FALSE}
sfsPoly <- t(apply(sfs[,-19], 1, nnorm))
barplot(sfsPoly, beside = TRUE, legend = c("AU", "NZ"), names = 1:18, main = "realSFS Polymorphic Sites")

barplot(sfs[c(-1, -length(sfs))])
barplot(AUsfs[c(-1, -length(AUsfs))])
```

## Fst
Here I plotted the Fst in sliding windows along chromosome 1.
```{r chromosome Fst, include=FALSE}
fst <- read.csv("angsd/chr1_fst_50KBwindows.tsv", sep = '\t', header = TRUE)

ggplot(fst, aes(x = fst, fill = replicate)) +
    geom_histogram() +
    theme_light()
mid <- mean(fst$fst)
fst_fig <- filter(fst, replicate == 'rep100') %>%
    ggplot(aes(x = pos, y = fst)) +
        geom_point(alpha = 0.25) +
        theme_light()
ggsave('plots/SNP_samtools_chr1_fst.png', fst_fig, device = "png", dpi = 600)
```