{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Diversity Indices\n",
    "### Runs of Homozygosity \n",
    "[ROHAN](https://github.com/grenaud/rohan)was used to call runs of homozygosity (ROHs). Following the recommendations, we removed all PCR-duplicates from BAM alignment files, estimated a transition/transversion ratio with [VCFtools](https://vcftools.github.io/index.html) v0.1.15, and defined an allowable level of heterozygosity within a 50kb window.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "# Update with how I estimate base-pair frequency and use that instead of TSTV ratio to estimate ROH."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And the Ts/Tv ratio was estimated as a prior for ROHan with VCFtools using the BCF outputs from ANGSD."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "for POP in AU TI\n",
    "    do\n",
    "    if [[ \"$POP\" == \"AU\" ]]\n",
    "        then\n",
    "        angsd -P 24 -b ${ANGSD}${POP}.list -ref $TREF -out ${ANGSD}samtools/genotypes/${POP}_genotypes \\\n",
    "            -uniqueOnly 1 -remove_bads 1 -only_proper_pairs 1 -trim 0 -C 50 -baq 1 \\\n",
    "            -minMapQ 20 -minQ 20 -minInd 19 -setMinDepth 200 -setMaxDepth 420 -doCounts 1 \\\n",
    "            -doPost 1 -doBcf 1 -GL 1 -doMajorMinor 1 -doMaf 1 -skipTriallelic 1 -SNP_pval 1e-6 -doGeno 10 --ignore-RG 0\n",
    "        vcftools --bcf ${ANGSD}samtools/genotypes/${POP}_genotypes.bcf \\\n",
    "            --TsTv-summary --out ${ANGSD}samtools/genotypes/${POP}_genotypes\n",
    "        elif [[ \"$POP\" == \"TI\" ]]\n",
    "        then\n",
    "        angsd -P 24 -b ${ANGSD}${POP}.list -ref $TREF -out ${ANGSD}samtools/genotypes/${POP}_genotypes \\\n",
    "            -uniqueOnly 1 -remove_bads 1 -only_proper_pairs 1 -trim 0 -C 50 -baq 1 \\\n",
    "            -minMapQ 20 -minQ 20 -minInd 15 -setMinDepth 120 -setMaxDepth 280 -doCounts 1 \\\n",
    "            -doPost 1 -doBcf 1 -GL 1 -doMajorMinor 1 -doMaf 1 -skipTriallelic 1 -SNP_pval 1e-6 -doGeno 10 --ignore-RG 0\n",
    "        vcftools --bcf ${ANGSD}samtools/genotypes/${POP}_genotypes.bcf \\\n",
    "            --TsTv-summary --out ${ANGSD}samtools/genotypes/${POP}_genotypes\n",
    "        else\n",
    "        angsd -P 24 -b ${ANGSD}${POP}.list -ref $KREF -out ${ANGSD}samtools/genotypes/${POP}_genotypes \\\n",
    "            -uniqueOnly 1 -remove_bads 1 -only_proper_pairs 1 -trim 0 -C 50 -baq 1 \\\n",
    "            -minMapQ 20 -minQ 20 -minInd 24 -setMinDepth 700 -setMaxDepth 1200 -doCounts 1 \\\n",
    "            -doPost 1 -doBcf 1 -GL 1 -doMajorMinor 1 -doMaf 1 -skipTriallelic 1 -SNP_pval 1e-3 -doGeno 10 --ignore-RG 0\n",
    "        vcftools --bcf ${ANGSD}samtools/genotypes/${POP}_genotypes.bcf \\\n",
    "            --TsTv-summary --out ${ANGSD}samtools/genotypes/${POP}_genotypes\n",
    "    fi\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This resulted in a TsTv ratio of 1.871 for tara iti, 2.507 for Australian fairy tern, and 2.725 for kakī.  \n",
    "\n",
    "Finally, ROHan was run with the `--rohmu` flag varied to ensure regions containing ROHs were detected as per [this discussion](https://github.com/grenaud/ROHan/issues/12#issuecomment-1935539239). The value of 5 x 10<sup>-5</sup> and a window size of 50kb equates 2.5 heterozygous genotypes within this window. For kakī, this threshold was left at the default setting of 1 x 10<sup>-5</sup> and equated to 0.5 heterozygous sites within a 50kb window. This much lower ROH threshold for kakī is attributable to the much higher sequence depth for the kakī data (i.e, target 50x vs 10x sequence depth).  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "./faSomeRecords reference/SP01_5kb_ragtag.fa SP01_autosomes2.bed reference/SP01_ragtag_autosomes2.fa\n",
    "./faSomeRecords reference/himNova-hic-scaff.fa himNova_autsosomes.bed reference/himNova_autosomes.fa\n",
    "\n",
    "samtools faidx reference/SP01_ragtag_autosomes2.fa\n",
    "samtools faidx reference/himNova_autosomes.fa\n",
    "\n",
    "REF=reference/SP01_ragtag_autosomes2.fa\n",
    "KIREF=reference/himNova_autosomes.fa\n",
    "\n",
    "for BAM in *_rohan.bam\n",
    "    do\n",
    "    BASE=$(basename $bam _rohan.bam)\n",
    "    printf \"STARTED RUNNING ROHAN FOR ${BASE} AT \"\n",
    "    date\n",
    "    if [[ \"$BASE\" == \"AU\"*]]\n",
    "        then\n",
    "        rohan -t 16 --tstv 2.507 --size 50000 --rohmu 5e-5 -o output/${BASE} $TREF $BAM\n",
    "    elif [[ \"$BASE\" == \"H0\"*]]\n",
    "        then\n",
    "        rohan -t 16 --tstv 2.725 --size 50000 -o output/${BASE} $KREF $BAM\n",
    "    else\n",
    "        rohan -t 16 --tstv 1.871 --size 50000 --rohmu 5e-5 -o output/${BASE} $TREF $BAM\n",
    "    fi\n",
    "    printf \"FINISHED RUNNING ROHAN FOR ${BASE} AT \"\n",
    "    date\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We then prepared summary plots for ROHs, one file each for tara iti and Australian fairy tern containing all ROH sizes calculated using the mid. estimates of heterozygosity.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "for FILE in fairy_tern/rohmu_5e5/AU*.mid.hmmrohl.gz\n",
    "    do\n",
    "    BASE=$(basename $FILE .mid.hmmrohl.gz)\n",
    "    zcat $FILE | tail -n+2 | cut -f 2-5 | awk -v SAMP=\"$BASE\" '{ if ( $4 <= 200000 ) len=\"Short ROH\";\n",
    "        else if ( $4 > 200000 && $4 <= 700000 ) len =\"Medium ROH\";\n",
    "        else len =\"Long ROH\";\n",
    "        print SAMP\"\\t\"$0\"\\t\"len\"\\tAU\"; }' >> ROHs.tsv\n",
    "done\n",
    "\n",
    "for FILE in fairy_tern/rohmu_5e5/{SND,SP,TI}*_subset.mid.hmmrohl.gz\n",
    "    do\n",
    "    BASE=$(basename $FILE _subset.mid.hmmrohl.gz)\n",
    "    zcat $FILE | tail -n+2 | cut -f 2-5 | awk -v SAMP=\"$BASE\" '{ if ( $4 <= 200000 ) len=\"Short ROH\";\n",
    "        else if ( $4 > 200000 && $4 <= 700000 ) len =\"Medium ROH\";\n",
    "        else len =\"Long ROH\";\n",
    "        print SAMP\"\\t\"$0\"\\t\"len\"\\tNZ\"; }' >> ROHs.tsv\n",
    "done\n",
    "\n",
    "for FILE in kaki/rohmu_default/H0*.mid.hmmrohl.gz\n",
    "    do\n",
    "    BASE=$(basename $FILE .mid.hmmrohl.gz)\n",
    "    zcat $FILE | tail -n+2 | cut -f 2-5 | awk -v SAMP=\"$BASE\" '{ if ( $4 <= 200000 ) len=\"Short ROH\";\n",
    "        else if ( $4 > 200000 && $4 <= 700000 ) len =\"Medium ROH\";\n",
    "        else len =\"Long ROH\";\n",
    "        print SAMP\"\\t\"$0\"\\t\"len\"\\tKI\"; }' >> ROHs.tsv\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We also parsed a file with subsetting the summary files of all individuals.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "printf \"Sample\\tPerc_in_ROH\\tMean_ROH Length\\tPopulation\\n\" > ROH_summary.txt\n",
    "\n",
    "for FILE in rohan/rohmu_5e5/*.summary.txt\n",
    "    do\n",
    "    BASE=$(basename $FILE .summary.txt)\n",
    "    PERC=$(tail -n4 $FILE | head -n1 | awk '{print $5}')\n",
    "    SIZE=$(tail -n1 $FILE | awk '{print $6}')\n",
    "    if [[ \"$BASE\" == \"AU\"* ]]\n",
    "        then\n",
    "        printf \"${BASE}\\t${PERC}\\t${SIZE}\\tAU\\n\" >> ROH_summary.txt\n",
    "    elif [[ \"$BASE\" == \"H0\"* ]]\n",
    "        then\n",
    "        printf \"${BASE}\\t${PERC}\\t${SIZE}\\tKI\\n\" >> ROH_summary.txt\n",
    "    else\n",
    "        printf \"${BASE}\\t${PERC}\\t${SIZE}\\tNZ\\n\" >> ROH_summary.txt\n",
    "    fi\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Visualising Inbreeding represented as *F<sub>ROH</sub>*\n",
    "Here we leverage ROH position and size estimates from ROHan to represent individual inbreeding the tara iti, AFT and kakī populations. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "c:\\Users\\jwold\\GitHub\\2024_MolEcol_ConsGen_Special_Issue\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "from matplotlib.legend_handler import HandlerTuple\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors as mcolors\n",
    "from scipy.stats import linregress\n",
    "from sklearn.metrics import r2_score\n",
    "from scipy.stats import mannwhitneyu\n",
    "from scipy import stats\n",
    "\n",
    "path = '/nesi/nobackup/uc03718/'\n",
    "os.chdir(path)\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We also examined the proportion of the genome impacted by ROHs for each individual of Australian fairy tern and tara iti. The higher count of ROHs in tara iti population also translated to a higher proportion of the genome impacted overall. However, the vast majority of ROHs impacting the genome is attributable to those in the 'large' (>700kb in length) category. This indicates that the tara iti population has remained chronically small for several generations.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "roh = pd.read_csv('ROHs.tsv', sep='\\t')\n",
    "roh_segments = pd.read_csv('ROH_segment_summary.txt', sep='\\t')\n",
    "\n",
    "roh_summary = roh[['Sample', 'Size Class', 'Population']]\n",
    "roh_summary = roh_summary.groupby(['Sample', 'Size Class', 'Population']).size().reset_index(name='ROH Counts')\n",
    "roh_summary = roh_summary.sort_values('Size Class', ascending=False)\n",
    "roh_summary['ROH_POP'] = roh_summary['Population'] + '-' + roh_summary['Size Class']\n",
    "\n",
    "roh_segments['Total Segments Included'] = roh_segments['Segments in ROH'] + roh_segments['Segments in non-ROH']\n",
    "\n",
    "roh_len = roh[['Sample', 'ROH Length', 'Size Class', 'Population']]\n",
    "roh_len = roh_len.groupby(['Sample', 'Size Class', 'Population'])['ROH Length'].sum().reset_index(name='ROH Total')\n",
    "\n",
    "\n",
    "roh_len = roh_len.merge(roh_segments[['Sample', 'Total Segments Included']], on='Sample', how='left')\n",
    "\n",
    "roh_len['Proportion'] = roh_len['ROH Total'] / roh_len['Total Segments Included']\n",
    "\n",
    "roh_len = roh_len.sort_values('Size Class', ascending=False)\n",
    "roh_len['ROH_POP'] = roh_len['Population'] + '-' + roh_len['Size Class']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we examined the ROH size distrobution. However, because medium and larger ROHs occur at a much lower frequency than small ROH, it is challenging to visualise the distribution. To help allevieate this, we log10 transformed ROH length prior to plotting.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Size Class</th>\n",
       "      <th>Population</th>\n",
       "      <th>ROH Total</th>\n",
       "      <th>Total Segments Included</th>\n",
       "      <th>Proportion</th>\n",
       "      <th>ROH_POP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>TI96</td>\n",
       "      <td>Short ROH</td>\n",
       "      <td>NZ</td>\n",
       "      <td>54900000</td>\n",
       "      <td>1133100000</td>\n",
       "      <td>0.048451</td>\n",
       "      <td>NZ-Short ROH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>TI85</td>\n",
       "      <td>Short ROH</td>\n",
       "      <td>NZ</td>\n",
       "      <td>61000000</td>\n",
       "      <td>1112150000</td>\n",
       "      <td>0.054849</td>\n",
       "      <td>NZ-Short ROH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>TI40</td>\n",
       "      <td>Short ROH</td>\n",
       "      <td>NZ</td>\n",
       "      <td>57250000</td>\n",
       "      <td>1069550000</td>\n",
       "      <td>0.053527</td>\n",
       "      <td>NZ-Short ROH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>TI38</td>\n",
       "      <td>Short ROH</td>\n",
       "      <td>NZ</td>\n",
       "      <td>68150000</td>\n",
       "      <td>1100200000</td>\n",
       "      <td>0.061943</td>\n",
       "      <td>NZ-Short ROH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>TI37</td>\n",
       "      <td>Short ROH</td>\n",
       "      <td>NZ</td>\n",
       "      <td>53850000</td>\n",
       "      <td>1111600000</td>\n",
       "      <td>0.048444</td>\n",
       "      <td>NZ-Short ROH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Sample Size Class Population  ROH Total  Total Segments Included  \\\n",
       "209   TI96  Short ROH         NZ   54900000               1133100000   \n",
       "176   TI85  Short ROH         NZ   61000000               1112150000   \n",
       "98    TI40  Short ROH         NZ   57250000               1069550000   \n",
       "95    TI38  Short ROH         NZ   68150000               1100200000   \n",
       "92    TI37  Short ROH         NZ   53850000               1111600000   \n",
       "\n",
       "     Proportion       ROH_POP  \n",
       "209    0.048451  NZ-Short ROH  \n",
       "176    0.054849  NZ-Short ROH  \n",
       "98     0.053527  NZ-Short ROH  \n",
       "95     0.061943  NZ-Short ROH  \n",
       "92     0.048444  NZ-Short ROH  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "roh_len[roh_len['PopulaNZon']=='NZ'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Population Diversity\n",
    "### Runs of Homozygosity (ROHs)\n",
    "We first examined the number of short (>50kb & <=200kb in length), medium (>200kb & <=700kb in length) and long (>700kb in length) ROHs per individual. This is to provide some indication of long-, medium- and short-term inbreeding levels respectively for each of the populations. Here, we can see that tara iti consistently have more ROHs across all three size classes.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mann-Whitney U Test results for ROH counts:\n",
      "AU-Short ROH vs AU-Medium ROH: U-statistic = 361.0, p-value = 1.4728853070303523e-07\n",
      "AU-Short ROH vs AU-Long ROH: U-statistic = 361.0, p-value = 1.4207436261002553e-07\n",
      "AU-Short ROH vs NZ-Short ROH: U-statistic = 24.0, p-value = 1.236627308196758e-09\n",
      "AU-Short ROH vs NZ-Medium ROH: U-statistic = 608.5, p-value = 0.10283822116259339\n",
      "AU-Short ROH vs NZ-Long ROH: U-statistic = 969.0, p-value = 1.6140058387966898e-10\n",
      "AU-Short ROH vs KI_10x-Short ROH: U-statistic = nan, p-value = nan\n",
      "AU-Short ROH vs KI_10x-Medium ROH: U-statistic = nan, p-value = nan\n",
      "AU-Short ROH vs KI_10x-Long ROH: U-statistic = nan, p-value = nan\n",
      "AU-Short ROH vs KI-Short ROH: U-statistic = nan, p-value = nan\n",
      "AU-Short ROH vs KI-Medium ROH: U-statistic = nan, p-value = nan\n",
      "AU-Short ROH vs KI-Long ROH: U-statistic = nan, p-value = nan\n",
      "AU-Medium ROH vs AU-Long ROH: U-statistic = 361.0, p-value = 1.418513185426635e-07\n",
      "AU-Medium ROH vs NZ-Short ROH: U-statistic = 0.0, p-value = 1.6324266271228804e-10\n",
      "AU-Medium ROH vs NZ-Medium ROH: U-statistic = 0.0, p-value = 1.6288465276339057e-10\n",
      "AU-Medium ROH vs NZ-Long ROH: U-statistic = 0.0, p-value = 1.6134147660288742e-10\n",
      "AU-Medium ROH vs KI_10x-Short ROH: U-statistic = nan, p-value = nan\n",
      "AU-Medium ROH vs KI_10x-Medium ROH: U-statistic = nan, p-value = nan\n",
      "AU-Medium ROH vs KI_10x-Long ROH: U-statistic = nan, p-value = nan\n",
      "AU-Medium ROH vs KI-Short ROH: U-statistic = nan, p-value = nan\n",
      "AU-Medium ROH vs KI-Medium ROH: U-statistic = nan, p-value = nan\n",
      "AU-Medium ROH vs KI-Long ROH: U-statistic = nan, p-value = nan\n",
      "AU-Long ROH vs NZ-Short ROH: U-statistic = 0.0, p-value = 1.6187414838975825e-10\n",
      "AU-Long ROH vs NZ-Medium ROH: U-statistic = 0.0, p-value = 1.615188572564675e-10\n",
      "AU-Long ROH vs NZ-Long ROH: U-statistic = 0.0, p-value = 1.5998740714897553e-10\n",
      "AU-Long ROH vs KI_10x-Short ROH: U-statistic = nan, p-value = nan\n",
      "AU-Long ROH vs KI_10x-Medium ROH: U-statistic = nan, p-value = nan\n",
      "AU-Long ROH vs KI_10x-Long ROH: U-statistic = nan, p-value = nan\n",
      "AU-Long ROH vs KI-Short ROH: U-statistic = nan, p-value = nan\n",
      "AU-Long ROH vs KI-Medium ROH: U-statistic = nan, p-value = nan\n",
      "AU-Long ROH vs KI-Long ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Short ROH vs NZ-Medium ROH: U-statistic = 2583.5, p-value = 9.182048171845107e-18\n",
      "NZ-Short ROH vs NZ-Long ROH: U-statistic = 2601.0, p-value = 3.27515256088414e-18\n",
      "NZ-Short ROH vs KI_10x-Short ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Short ROH vs KI_10x-Medium ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Short ROH vs KI_10x-Long ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Short ROH vs KI-Short ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Short ROH vs KI-Medium ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Short ROH vs KI-Long ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Medium ROH vs NZ-Long ROH: U-statistic = 2601.0, p-value = 3.2708934204205383e-18\n",
      "NZ-Medium ROH vs KI_10x-Short ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Medium ROH vs KI_10x-Medium ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Medium ROH vs KI_10x-Long ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Medium ROH vs KI-Short ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Medium ROH vs KI-Medium ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Medium ROH vs KI-Long ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Long ROH vs KI_10x-Short ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Long ROH vs KI_10x-Medium ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Long ROH vs KI_10x-Long ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Long ROH vs KI-Short ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Long ROH vs KI-Medium ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Long ROH vs KI-Long ROH: U-statistic = nan, p-value = nan\n",
      "KI_10x-Short ROH vs KI_10x-Medium ROH: U-statistic = nan, p-value = nan\n",
      "KI_10x-Short ROH vs KI_10x-Long ROH: U-statistic = nan, p-value = nan\n",
      "KI_10x-Short ROH vs KI-Short ROH: U-statistic = nan, p-value = nan\n",
      "KI_10x-Short ROH vs KI-Medium ROH: U-statistic = nan, p-value = nan\n",
      "KI_10x-Short ROH vs KI-Long ROH: U-statistic = nan, p-value = nan\n",
      "KI_10x-Medium ROH vs KI_10x-Long ROH: U-statistic = nan, p-value = nan\n",
      "KI_10x-Medium ROH vs KI-Short ROH: U-statistic = nan, p-value = nan\n",
      "KI_10x-Medium ROH vs KI-Medium ROH: U-statistic = nan, p-value = nan\n",
      "KI_10x-Medium ROH vs KI-Long ROH: U-statistic = nan, p-value = nan\n",
      "KI_10x-Long ROH vs KI-Short ROH: U-statistic = nan, p-value = nan\n",
      "KI_10x-Long ROH vs KI-Medium ROH: U-statistic = nan, p-value = nan\n",
      "KI_10x-Long ROH vs KI-Long ROH: U-statistic = nan, p-value = nan\n",
      "KI-Short ROH vs KI-Medium ROH: U-statistic = nan, p-value = nan\n",
      "KI-Short ROH vs KI-Long ROH: U-statistic = nan, p-value = nan\n",
      "KI-Medium ROH vs KI-Long ROH: U-statistic = nan, p-value = nan\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\jwold\\AppData\\Local\\Temp\\ipykernel_12048\\556418944.py:4: SmallSampleWarning: One or more sample arguments is too small; all returned values will be NaN. See documentation for sample size requirements.\n",
      "  stat, p_value = mannwhitneyu(group1_values, group2_values, alternative='two-sided')\n"
     ]
    }
   ],
   "source": [
    "def mann_whitney_test(df, group_col, value_col, group1, group2):\n",
    "    group1_values = df[df[group_col] == group1][value_col]\n",
    "    group2_values = df[df[group_col] == group2][value_col]\n",
    "    stat, p_value = mannwhitneyu(group1_values, group2_values, alternative='two-sided')\n",
    "    return stat, p_value\n",
    "\n",
    "# Perform the test for indiv_total\n",
    "groups = ['AU-Short ROH', 'AU-Medium ROH', 'AU-Long ROH', 'NZ-Short ROH', 'NZ-Medium ROH', 'NZ-Long ROH', 'KI_10x-Short ROH', 'KI_10x-Medium ROH', 'KI_10x-Long ROH', 'KI-Short ROH', 'KI-Medium ROH', 'KI-Long ROH']\n",
    "results_total = {}\n",
    "\n",
    "for i in range(len(groups)):\n",
    "    for j in range(i + 1, len(groups)):\n",
    "        group1 = groups[i]\n",
    "        group2 = groups[j]\n",
    "        stat, p_value = mann_whitney_test(roh_summary, 'ROH_POP', 'ROH Counts', group1, group2)\n",
    "        results_total[f'{group1} vs {group2}'] = {'stat': stat, 'p_value': p_value}\n",
    "\n",
    "# Print the results\n",
    "print(\"Mann-Whitney U Test results for ROH counts:\")\n",
    "for comparison, result in results_total.items():\n",
    "    print(f\"{comparison}: U-statistic = {result['stat']}, p-value = {result['p_value']}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Segments Unclassified</th>\n",
       "      <th>Segments in ROH</th>\n",
       "      <th>Segments in non-ROH</th>\n",
       "      <th>Population</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AU01</td>\n",
       "      <td>83900000</td>\n",
       "      <td>99700000</td>\n",
       "      <td>1046000000</td>\n",
       "      <td>AU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AU03</td>\n",
       "      <td>83900000</td>\n",
       "      <td>119650000</td>\n",
       "      <td>1026050000</td>\n",
       "      <td>AU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AU04</td>\n",
       "      <td>79200000</td>\n",
       "      <td>77050000</td>\n",
       "      <td>1073350000</td>\n",
       "      <td>AU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AU06</td>\n",
       "      <td>86700000</td>\n",
       "      <td>83000000</td>\n",
       "      <td>1059900000</td>\n",
       "      <td>AU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AU08</td>\n",
       "      <td>72350000</td>\n",
       "      <td>82850000</td>\n",
       "      <td>1074400000</td>\n",
       "      <td>AU</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample  Segments Unclassified  Segments in ROH  Segments in non-ROH  \\\n",
       "0   AU01               83900000         99700000           1046000000   \n",
       "1   AU03               83900000        119650000           1026050000   \n",
       "2   AU04               79200000         77050000           1073350000   \n",
       "3   AU06               86700000         83000000           1059900000   \n",
       "4   AU08               72350000         82850000           1074400000   \n",
       "\n",
       "  Population  \n",
       "0         AU  \n",
       "1         AU  \n",
       "2         AU  \n",
       "3         AU  \n",
       "4         AU  "
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "roh_segments.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mean individual FROH for AU: 0.07738820920883634 +/- 0.0025104285507874502\n",
      "Mean individual FROH for NZ: 0.6610220793959886 +/- 0.006279098298593244\n",
      "Mean individual FROH for KI_10x: nan +/- nan\n",
      "Mean individual FROH for KI: nan +/- nan\n"
     ]
    }
   ],
   "source": [
    "roh_proportion = roh.groupby(['Sample', 'Population'])['ROH Length'].sum().reset_index(name='ROH Total')\n",
    "indiv_roh_counts = roh.groupby(['Sample', 'Population'])['ROH Length'].count().reset_index(name='Count').sort_values('Count')\n",
    "roh_segments = pd.read_csv('ROH_segment_summary.txt', sep='\\t')\n",
    "roh_segments['Total Segments Included'] = roh_segments['Segments in ROH'] + roh_segments['Segments in non-ROH']\n",
    "\n",
    "roh_proportion = roh_proportion.merge(roh_segments[['Sample', 'Total Segments Included']], on='Sample', how='left')\n",
    "roh_proportion['Froh'] = roh_proportion['ROH Total'] / roh_proportion['Total Segments Included']\n",
    "\n",
    "roh_proportion.sort_values('Froh')\n",
    "\n",
    "groups = ['AU', 'NZ', 'KI_10x', 'KI']\n",
    "\n",
    "for i in range(len(groups)):\n",
    "    standard_dev = roh_proportion[roh_proportion['Population']==groups[i]]['Froh'].std()\n",
    "    count = roh_proportion[roh_proportion['Population']==groups[i]]['Froh'].count()\n",
    "    standard_err = standard_dev / np.sqrt(count)\n",
    "    print(f\"Mean individual FROH for {groups[i]}: {roh_proportion[roh_proportion['Population']==groups[i]]['Froh'].mean()} +/- {standard_err}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mann-Whitney U Test results for FROH:\n",
      "AU-Short ROH vs AU-Medium ROH: U-statistic = 322.0, p-value = 3.8473714744244065e-05\n",
      "AU-Short ROH vs AU-Long ROH: U-statistic = 341.0, p-value = 2.995041856121807e-06\n",
      "AU-Short ROH vs NZ-Short ROH: U-statistic = 4.0, p-value = 2.3082251058670538e-10\n",
      "AU-Short ROH vs NZ-Medium ROH: U-statistic = 0.0, p-value = 1.6360138495636074e-10\n",
      "AU-Short ROH vs NZ-Long ROH: U-statistic = 0.0, p-value = 1.6360138495636074e-10\n",
      "AU-Short ROH vs KI_10x-Short ROH: U-statistic = nan, p-value = nan\n",
      "AU-Short ROH vs KI_10x-Medium ROH: U-statistic = nan, p-value = nan\n",
      "AU-Short ROH vs KI_10x-Long ROH: U-statistic = nan, p-value = nan\n",
      "AU-Short ROH vs KI-Short ROH: U-statistic = nan, p-value = nan\n",
      "AU-Short ROH vs KI-Medium ROH: U-statistic = nan, p-value = nan\n",
      "AU-Short ROH vs KI-Long ROH: U-statistic = nan, p-value = nan\n",
      "AU-Medium ROH vs AU-Long ROH: U-statistic = 327.0, p-value = 2.0221305879843327e-05\n",
      "AU-Medium ROH vs NZ-Short ROH: U-statistic = 0.0, p-value = 1.6360138495636074e-10\n",
      "AU-Medium ROH vs NZ-Medium ROH: U-statistic = 0.0, p-value = 1.6360138495636074e-10\n",
      "AU-Medium ROH vs NZ-Long ROH: U-statistic = 0.0, p-value = 1.6360138495636074e-10\n",
      "AU-Medium ROH vs KI_10x-Short ROH: U-statistic = nan, p-value = nan\n",
      "AU-Medium ROH vs KI_10x-Medium ROH: U-statistic = nan, p-value = nan\n",
      "AU-Medium ROH vs KI_10x-Long ROH: U-statistic = nan, p-value = nan\n",
      "AU-Medium ROH vs KI-Short ROH: U-statistic = nan, p-value = nan\n",
      "AU-Medium ROH vs KI-Medium ROH: U-statistic = nan, p-value = nan\n",
      "AU-Medium ROH vs KI-Long ROH: U-statistic = nan, p-value = nan\n",
      "AU-Long ROH vs NZ-Short ROH: U-statistic = 0.0, p-value = 1.6360138495636074e-10\n",
      "AU-Long ROH vs NZ-Medium ROH: U-statistic = 0.0, p-value = 1.6360138495636074e-10\n",
      "AU-Long ROH vs NZ-Long ROH: U-statistic = 0.0, p-value = 1.6360138495636074e-10\n",
      "AU-Long ROH vs KI_10x-Short ROH: U-statistic = nan, p-value = nan\n",
      "AU-Long ROH vs KI_10x-Medium ROH: U-statistic = nan, p-value = nan\n",
      "AU-Long ROH vs KI_10x-Long ROH: U-statistic = nan, p-value = nan\n",
      "AU-Long ROH vs KI-Short ROH: U-statistic = nan, p-value = nan\n",
      "AU-Long ROH vs KI-Medium ROH: U-statistic = nan, p-value = nan\n",
      "AU-Long ROH vs KI-Long ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Short ROH vs NZ-Medium ROH: U-statistic = 16.0, p-value = 8.440634793530077e-18\n",
      "NZ-Short ROH vs NZ-Long ROH: U-statistic = 0.0, p-value = 3.3036815016661564e-18\n",
      "NZ-Short ROH vs KI_10x-Short ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Short ROH vs KI_10x-Medium ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Short ROH vs KI_10x-Long ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Short ROH vs KI-Short ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Short ROH vs KI-Medium ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Short ROH vs KI-Long ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Medium ROH vs NZ-Long ROH: U-statistic = 0.0, p-value = 3.3036815016661564e-18\n",
      "NZ-Medium ROH vs KI_10x-Short ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Medium ROH vs KI_10x-Medium ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Medium ROH vs KI_10x-Long ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Medium ROH vs KI-Short ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Medium ROH vs KI-Medium ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Medium ROH vs KI-Long ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Long ROH vs KI_10x-Short ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Long ROH vs KI_10x-Medium ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Long ROH vs KI_10x-Long ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Long ROH vs KI-Short ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Long ROH vs KI-Medium ROH: U-statistic = nan, p-value = nan\n",
      "NZ-Long ROH vs KI-Long ROH: U-statistic = nan, p-value = nan\n",
      "KI_10x-Short ROH vs KI_10x-Medium ROH: U-statistic = nan, p-value = nan\n",
      "KI_10x-Short ROH vs KI_10x-Long ROH: U-statistic = nan, p-value = nan\n",
      "KI_10x-Short ROH vs KI-Short ROH: U-statistic = nan, p-value = nan\n",
      "KI_10x-Short ROH vs KI-Medium ROH: U-statistic = nan, p-value = nan\n",
      "KI_10x-Short ROH vs KI-Long ROH: U-statistic = nan, p-value = nan\n",
      "KI_10x-Medium ROH vs KI_10x-Long ROH: U-statistic = nan, p-value = nan\n",
      "KI_10x-Medium ROH vs KI-Short ROH: U-statistic = nan, p-value = nan\n",
      "KI_10x-Medium ROH vs KI-Medium ROH: U-statistic = nan, p-value = nan\n",
      "KI_10x-Medium ROH vs KI-Long ROH: U-statistic = nan, p-value = nan\n",
      "KI_10x-Long ROH vs KI-Short ROH: U-statistic = nan, p-value = nan\n",
      "KI_10x-Long ROH vs KI-Medium ROH: U-statistic = nan, p-value = nan\n",
      "KI_10x-Long ROH vs KI-Long ROH: U-statistic = nan, p-value = nan\n",
      "KI-Short ROH vs KI-Medium ROH: U-statistic = nan, p-value = nan\n",
      "KI-Short ROH vs KI-Long ROH: U-statistic = nan, p-value = nan\n",
      "KI-Medium ROH vs KI-Long ROH: U-statistic = nan, p-value = nan\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\jwold\\AppData\\Local\\Temp\\ipykernel_12048\\556418944.py:4: SmallSampleWarning: One or more sample arguments is too small; all returned values will be NaN. See documentation for sample size requirements.\n",
      "  stat, p_value = mannwhitneyu(group1_values, group2_values, alternative='two-sided')\n"
     ]
    }
   ],
   "source": [
    "groups = ['AU-Short ROH', 'AU-Medium ROH', 'AU-Long ROH', 'NZ-Short ROH', 'NZ-Medium ROH', 'NZ-Long ROH', 'KI_10x-Short ROH', 'KI_10x-Medium ROH', 'KI_10x-Long ROH', 'KI-Short ROH', 'KI-Medium ROH', 'KI-Long ROH']\n",
    "results_total = {}\n",
    "\n",
    "for i in range(len(groups)):\n",
    "    for j in range(i + 1, len(groups)):\n",
    "        group1 = groups[i]\n",
    "        group2 = groups[j]\n",
    "        stat, p_value = mann_whitney_test(roh_len, 'ROH_POP', 'Proportion', group1, group2)\n",
    "        results_total[f'{group1} vs {group2}'] = {'stat': stat, 'p_value': p_value}\n",
    "\n",
    "# Print the results\n",
    "print(\"Mann-Whitney U Test results for FROH:\")\n",
    "for comparison, result in results_total.items():\n",
    "    print(f\"{comparison}: U-statistic = {result['stat']}, p-value = {result['p_value']}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Sample Population  ROH Total  Total Segments Included      Froh\n",
      "27   TI21         NZ  513850000               1024550000  0.501537\n",
      "21  SND11         NZ  589450000               1047750000  0.562586\n",
      "20  SND06         NZ  610650000               1059850000  0.576166\n",
      "19  SND04         NZ  617750000               1059100000  0.583278\n",
      "43   TI65         NZ  641150000               1071550000  0.598339\n",
      "32   TI40         NZ  648400000               1069550000  0.606236\n",
      "28   TI35         NZ  657300000               1081850000  0.607570\n",
      "55   TI82         NZ  678450000               1086850000  0.624235\n",
      "56   TI83         NZ  671350000               1074750000  0.624657\n",
      "23   SP03         NZ  674900000               1078350000  0.625864\n",
      "57   TI84         NZ  683850000               1076950000  0.634988\n",
      "53   TI78         NZ  692950000               1089200000  0.636201\n",
      "48   TI70         NZ  697850000               1093300000  0.638297\n",
      "40   TI62         NZ  698050000               1089250000  0.640854\n",
      "25  TI101         NZ  703650000               1097100000  0.641373\n",
      "61   TI88         NZ  699250000               1086500000  0.643580\n",
      "31   TI38         NZ  711400000               1100200000  0.646610\n",
      "47   TI69         NZ  709250000               1089250000  0.651136\n",
      "46   TI68         NZ  723250000               1101750000  0.656456\n",
      "36   TI52         NZ  710600000               1082200000  0.656625\n",
      "33   TI41         NZ  724800000               1102300000  0.657534\n",
      "63   TI90         NZ  728200000               1103250000  0.660050\n",
      "29   TI36         NZ  720000000               1089950000  0.660581\n",
      "22   SP02         NZ  730850000               1103950000  0.662032\n",
      "37   TI54         NZ  723650000               1092450000  0.662410\n",
      "34   TI50         NZ  740800000               1103100000  0.671562\n",
      "52   TI77         NZ  743700000               1105850000  0.672514\n",
      "38   TI56         NZ  744050000               1104550000  0.673623\n",
      "30   TI37         NZ  749400000               1111600000  0.674163\n",
      "64   TI91         NZ  745250000               1102800000  0.675780\n",
      "60   TI87         NZ  745550000               1099350000  0.678173\n",
      "41   TI63         NZ  747650000               1102250000  0.678294\n",
      "49   TI71         NZ  760150000               1111500000  0.683896\n",
      "24  TI100         NZ  761250000               1108950000  0.686460\n",
      "26  TI102         NZ  766900000               1116950000  0.686602\n",
      "42   TI64         NZ  762200000               1109200000  0.687162\n",
      "45   TI67         NZ  771350000               1114100000  0.692353\n",
      "51   TI75         NZ  771650000               1112350000  0.693712\n",
      "44   TI66         NZ  776000000               1117050000  0.694687\n",
      "58   TI85         NZ  775750000               1112150000  0.697523\n",
      "39   TI57         NZ  781850000               1120350000  0.697862\n",
      "62   TI89         NZ  783200000               1120550000  0.698942\n",
      "54   TI81         NZ  789850000               1125350000  0.701871\n",
      "68   TI95         NZ  783100000               1113200000  0.703467\n",
      "69   TI96         NZ  802550000               1133100000  0.708278\n",
      "50   TI72         NZ  804500000               1133800000  0.709561\n",
      "66   TI93         NZ  799350000               1123150000  0.711704\n",
      "59   TI86         NZ  787600000               1103000000  0.714053\n",
      "67   TI94         NZ  811050000               1130350000  0.717521\n",
      "35   TI51         NZ  812200000               1131550000  0.717777\n",
      "65   TI92         NZ  827200000               1140350000  0.725391\n"
     ]
    }
   ],
   "source": [
    "print(roh_proportion[roh_proportion['Population']=='NZ'].sort_values(by='Froh', ascending=True))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The individuals with the smallest, median, and largest *F<sub>ROH</sub>* for each group.  \n",
    "- Australian fairy tern\n",
    "    - Smallest *F<sub>ROH</sub>*: AU13 (3.036%)\n",
    "    - Median *F<sub>ROH</sub>*: AU08 (3.890%)\n",
    "    - Largest *F<sub>ROH</sub>*: AU03 (6.801%)\n",
    "- Tara iti\n",
    "    - Smallest *F<sub>ROH</sub>*: TI21 (27.149%)\n",
    "    - Median *F<sub>ROH</sub>*: SP07 (42.051%)\n",
    "    - Largest *F<sub>ROH</sub>*: SND05 (59.947%)\n",
    "- Kakī low-coverage\n",
    "    - Smallest *F<sub>ROH</sub>*: H01402 (24.625%)\n",
    "    - Median *F<sub>ROH</sub>*: H01406 (36.51%)\n",
    "    - Largest *F<sub>ROH</sub>*: H01387 (49.57%)\n",
    "- Kakī high-coverage\n",
    "    - Smallest *F<sub>ROH</sub>*: H01396 (17.816%)\n",
    "    - Median *F<sub>ROH</sub>*: H01406 (31.092%)\n",
    "    - Largest *F<sub>ROH</sub>*: H01390 (38.175%)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "roh_plot = pd.read_csv('ROHs.tsv', sep='\\t')\n",
    "roh_plot['IDs'] = roh_plot['Sample'] + '-' + roh_plot['Population']\n",
    "\n",
    "chrom = 'Chromosome 1'\n",
    "individuals = [\n",
    "    'AU13-AU', 'AU08-AU', 'AU03-AU',\n",
    "    'TI21-TI', 'TI40-TI', 'SND05-TI',\n",
    "    'H01401-KI_10x', 'H01386-KI_10x', 'H01394-KI_10x',\n",
    "    'H01396-KI', 'H01406-KI', 'H01390-KI'\n",
    "    ]\n",
    "populations = [\n",
    "    'AFT1', 'AFT2', 'AFT3',\n",
    "    'TI1', 'TI2', 'TI3',\n",
    "    'KĪ low1', 'KĪ low2', 'KĪ low3',\n",
    "    'KĪ1', 'KĪ2', 'KĪ3'\n",
    "    ]\n",
    "\n",
    "subset_roh = roh[roh['Sample'].isin(individuals)]\n",
    "subset_roh = subset_roh[(subset_roh['Chromosome'] == 'CM020437.1_RagTag') | (subset_roh['Chromosome'] == 'scaffold_1')]\n",
    "subset_roh['Chromosome'] = chrom\n",
    "subset_roh = subset_roh.sort_values(by=['Sample', 'BEGIN'])\n",
    "\n",
    "plot_rohs(subset_roh, chrom, individuals, populations, chr1_size, color_mapping=color_mapping, figsize=(20, 10))\n",
    "plt.savefig('plots/ROH_positions.png', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We then examined the density of Runs of Homozygosity (ROHs) by size class. Here, we plot ROHs presence at a particular location across individuals for chromosome 1 as a means to visualise relative *fROH* among individuals.  \n",
    "\n",
    "To start, we load and subset the original `ROH_density.tsv` file by population, ROH size class, and position (`BEGIN` and `END`).  \n",
    "\n",
    "We then define a plot function that marks the presence/absence of ROHs by their `BEGIN` and `END` position along chromosome 1. The color `seagreen` denotes ROHs smaller than or equal to 200kb in length, `wheat` indicates ROHs larger than 200kb and smaller than or equal to 700kb, and `coral` represents ROHs larger than 700kb.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace with code from kakakpo\n",
    "\n",
    "def plot_rohs(df, chromosome, individuals, populations, continuous_range, alpha_multiplier=1, color_mapping=None, figsize=(20, 15)):\n",
    "    \"\"\"\n",
    "    Plot ROH presence along the chromosome for specific individuals.\n",
    "\n",
    "    Parameters:\n",
    "        df (DataFrame): Input DataFrame containing 'Chromosome', 'BEGIN', 'END', 'Size Class', 'Sample', and 'Population' columns.\n",
    "        chromosome (str): Chromosome to plot.\n",
    "        individuals (list): List of individuals to include in the plot.\n",
    "        populations (list): List of populations corresponding to the individuals.\n",
    "        continuous_range (range or list): Continuous range of positions to consider.\n",
    "        alpha_multiplier (float): Multiplier to adjust alpha based on the number of overlapping ranges.\n",
    "        color_mapping (dict): Mapping of ROH classes to colors.\n",
    "        figsize (tuple, optional): Size of the figure in inches (width, height).\n",
    "\n",
    "    Returns:\n",
    "        None\n",
    "    \"\"\"\n",
    "    num_individuals = len(individuals)\n",
    "    fig, axs = plt.subplots(num_individuals, 1, figsize=figsize, sharex=True, sharey=True)\n",
    "\n",
    "    for i, (individual, population) in enumerate(zip(individuals, populations)):\n",
    "        ax = axs[i]\n",
    "        # Filter Pandas DataFrame for the specified chromosome and individual\n",
    "        chrom_ind_df = df[(df['Chromosome'] == chromosome) & (df['IDs'] == individual)]\n",
    "        \n",
    "        # Get color for ROH class\n",
    "        if color_mapping:\n",
    "            chrom_ind_df['Color'] = chrom_ind_df['Size Class'].map(color_mapping)\n",
    "        else:\n",
    "            chrom_ind_df['Color'] = 'steelblue'  # Default color\n",
    "\n",
    "        # Plot ROH presence\n",
    "        for _, row in chrom_ind_df.iterrows():\n",
    "            start, end = row['BEGIN'], row['END']\n",
    "            ax.fill_between(range(start, end+1), 0, 1, color=row['Color'], alpha=alpha_multiplier, edgecolor=None)\n",
    "        \n",
    "        # Set y-axis label to population\n",
    "        ax.set_ylabel(population, labelpad=5, fontsize=18)\n",
    "\n",
    "        # Remove y-axis ticks\n",
    "        ax.set_xticks([])\n",
    "        ax.set_yticks([])\n",
    "\n",
    "        # Set x-axis limits\n",
    "        ax.set_xlim(min(continuous_range), max(continuous_range))\n",
    "        ax.set_ylim(0, 1)\n",
    "\n",
    "    axs[-1].set_xlabel('Position on Chromosome 1', fontsize=25)\n",
    "    axs[0].set_title('E)', loc='left', fontsize=35)\n",
    "    plt.tight_layout()\n",
    "\n",
    "# Define Color mapping\n",
    "color_mapping = {\n",
    "    'Short ROH': 'seagreen',\n",
    "    'Medium ROH': 'wheat',\n",
    "    'Long ROH': 'coral'\n",
    "}\n",
    "# Define chromosome size\n",
    "chr1_size = range(0, 219000000, 1000)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Individual Heterozygosity (H<sub>o</sub>)\n",
    "#### SNPs\n",
    "Here, we implemented a global (genome-wide heterozygosity) method from ANGSD. Essentially, this estimate is a proportion of heterozygous genotypes / genome size (excluding regions of the genome with low confidence). Unlike other runs of ANGSD, individual BAMs are used to estimate hetereozygosity, which is simply second value in the SFS/AFS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "SAMP=$(sed -n \"$((SLURM_ARRAY_TASK_ID + 1))p\" /nesi/nobackup/uc03718/angsd/GLOBAL.list)\n",
    "NAME=$(basename $SAMP _autosomes_nodup.bam)\n",
    "\n",
    "printf \"\\nSTARTED RUNNING ANGSD TO ESTIMATE HETEROZYGOSITY FOR $NAME AT \"\n",
    "date\n",
    "\n",
    "angsd -P 32 -i ${SAMP} -ref $TREF -out ${DIR}samtools/heterozygosity/${NAME} \\\n",
    "        -uniqueOnly 1 -remove_bads 1 -only_proper_pairs 1 -trim 0 -C 50 -angbaq 1 \\\n",
    "        -minMapQ 20 -minQ 20 -doCounts 1 -dosaf 1 -GL 1\n",
    "\n",
    "realSFS ${DIR}samtools/heterozygosity/${NAME}.saf.ids > ${DIR}samtools/heterozygosity/${NAME}_est.ml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once the SFS was estimated for each individual, the number of sites was estimated from the sum of all scaffold sizes included in the bam file and output to a file.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "printf \"Sample\\tHeterozygosity\\tTool\\tPopulation\\n\" > ${ANGSD}individual_het.tsv \n",
    "\n",
    "for SAMP in ${ANGSD}${TOOL}/heterozygosity/*_est.ml\n",
    "    do\n",
    "    BASE=$(basename $SAMP _est.ml)\n",
    "    TOT=$(awk '{print $1 + $2 + $3}' $SAMP)\n",
    "    HET=$(awk -v var=$TOT '{print $2/var}' $SAMP)\n",
    "    if [[ \"$BASE\" == \"AU\"* ]]\n",
    "        then\n",
    "        printf \"$BASE\\t$HET\\t$TOOL\\tAU\\n\" >> ${ANGSD}individual_het.tsv\n",
    "    elif [[ \"$BASE\" == \"H0\"* ]]\n",
    "        then\n",
    "        printf \"$BASE\\t$HET\\t$TOOL\\tKI\\n\" >> ${ANGSD}individual_het.tsv\n",
    "    else\n",
    "        printf \"$BASE\\t$HET\\t$TOOL\\tNZ\\n\" >> ${ANGSD}individual_het.tsv\n",
    "    fi\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### SVs\n",
    "Genotypes for each individual were estimated and then plotted in 06_visualisations.ipynb.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "while read -r line\n",
    "    do\n",
    "    REF=$(bcftools query -s ${line} -i 'GT!=\"mis\" & GQ>=25 & SVMODEL==\"AGGREGATED\" & FORMAT/FT==\"PASS\" & GT==\"RR\"' -f '[%GT\\n]' graphtyper/02_fairy_genotypes_filtered.vcf.gz | wc -l)\n",
    "    HET=$(bcftools query -s ${line} -i 'GT!=\"mis\" & GQ>=25 & SVMODEL==\"AGGREGATED\" & FORMAT/FT==\"PASS\" & GT==\"het\"' -f '[%GT\\n]' graphtyper/02_fairy_genotypes_filtered.vcf.gz | wc -l | awk -v var=$TOT '{print $1/var}')\n",
    "    ALT=$(bcftools query -s ${line} -i 'GT!=\"mis\" & GQ>=25 & SVMODEL==\"AGGREGATED\" & FORMAT/FT==\"PASS\" & GT==\"AA\"' -f '[%GT\\n]' graphtyper/02_fairy_genotypes_filtered.vcf.gz | wc -l)\n",
    "    if [[ \"$line\" == \"AU\"* ]]\n",
    "    then\n",
    "        printf \"$line\\t$REF\\t$HET\\t$ALT\\tAU\\n\"\n",
    "        printf \"$line\\t$REF\\t$HET\\t$ALT\\tAU\\n\" >> graphtyper/individual_svHet.tsv\n",
    "    else\n",
    "        printf \"$line\\t$REF\\t$HET\\t$ALT\\tTI\\n\"\n",
    "        printf \"$line\\t$REF\\t$HET\\t$ALT\\tTI\\n\" >> graphtyper/individual_svHet.tsv\n",
    "    fi\n",
    "done < samples.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SNP- and SV-based *H<sub>O</sub>* Estimates\n",
    "The next diversity metric we examined was individual observed heterozygosity. SNPs were estimated using `ANGSD` and `realSFS` for all sites and putatively neutral sites where we excluded regions predicted to be coding regions by `AUGUSTUS`.  \n",
    "\n",
    "We estimated structural variant heterozygosity as:\n",
    "$$\n",
    "H_{O} = \\frac{\\sum{H Sites}}{\\sum{Genotyped Sites}}\n",
    "$$\n",
    "\n",
    "To examine the impact of read depth on SNPs and SVs, we leveraged the heterozygosity estimated using Australian fairy tern, tara iti, kakī subsampled alignments and high coverage kakī alignments. For esimtates of SV heterozygosity, we filtered out all invariant sites in the Australian fairy tern and tara iti populations. This included fixed SVs in Australian fairy terns.   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sv_het = pd.read_csv('graphtyper/individual_svHet.tsv', delimiter='\\t')\n",
    "sv_het = sv_het[\n",
    "                (sv_het['Population']!='50xSVs_10xGTs') & \n",
    "                (sv_het['Population']!='50xSVs_50xGTs') & \n",
    "                (sv_het['Population']!='AU_PlotCritic') &\n",
    "                (sv_het['Population']!='AU_sites') & \n",
    "                (sv_het['Population']!='AU') & \n",
    "                (sv_het['Population']!='TI_PlotCritic') & \n",
    "                (sv_het['Population']!='TI')\n",
    "                ]\n",
    "\n",
    "print(sv_het[sv_het['Population']=='KI'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Kakī SV Mendelian Inheritance\n",
    "Here we examine the number of sites adhering to mendelian inheritance expectations in X trios for 3 data sets filtered for genotyping quality:\n",
    "1) SVs discovered and genotyped using high coverage data\n",
    "2) SVs discovered using high coverage data, and genotyped using subset data\n",
    "3) SVs discovered and genotyped using subset data\n",
    "\n",
    "Here, high coverage data is ranging from ~22x - ~67x alignment depth while subset data represents aligned depths ranging from ~6x = ~13x.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kiSVmendel = pd.read_csv('graphtyper/graphtyper_kaki_subset/KI_mendelian_counts.tsv', sep='\\t')\n",
    "\n",
    "kiSVmendel.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kiSVmendel.groupby(['Data set'])['Percent Failing'].mean().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, we are briefly examining the average proportion of SVs in either a homozygous reference or alternate state in Australian fairy tern, tara iti, and subsampled kakī data sets.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sv_het = pd.read_csv('graphtyper/individual_svHet.tsv', delimiter='\\t')\n",
    "kiSVmendel = kiSVmendel[kiSVmendel['Data set']!='50xSVs_10xGenotypes']\n",
    "\n",
    "order = ['KI_10x', '50xSVs_50xGenotypes']\n",
    "\n",
    "# Grayscale colors for the three categories\n",
    "colors = ['#d9d9d9', '#969696', '#525252']  # Light, medium, and dark gray\n",
    "\n",
    "# Create subplots\n",
    "fig, axes = plt.subplots(3, 2, figsize=(20, 20), sharex=False, sharey=False)  # 1 row, 4 columns, shared y-axis\n",
    "\n",
    "au_svHet = sv_het[sv_het['Population'] == 'AU_noFixed']  # Filter for the current population\n",
    "ti_svHet = sv_het[sv_het['Population'] == 'TI_sites']\n",
    "kiLC_svHet = sv_het[sv_het['Population'] == 'KI_10x']\n",
    "kiHC_svHet = sv_het[sv_het['Population'] == 'KI']\n",
    "aux = np.arange(len(au_svHet))\n",
    "tix = np.arange(len(ti_svHet))\n",
    "kix = np.arange(len(kiLC_svHet))\n",
    "\n",
    "# Plot stacked bars\n",
    "axes[0,0].bar(aux, au_svHet['Homozygous Ref'], label='Homozygous Ref', color=colors[0])\n",
    "axes[0,0].bar(aux, au_svHet['Heterozygous'], bottom=au_svHet['Homozygous Ref'], label='Heterozygous', color=colors[1])\n",
    "axes[0,0].bar(aux, au_svHet['Homozygous Alt'], bottom=au_svHet['Homozygous Ref'] + au_svHet['Heterozygous'], label='Homozygous Alt', color=colors[2])\n",
    "axes[0,0].set_xticks([])\n",
    "axes[0,0].set_title('A)', fontsize = 20, loc='left')\n",
    "axes[0,0].set_ylim(0,1800)\n",
    "\n",
    "axes[0,1].bar(tix, ti_svHet['Homozygous Ref'], label='Homozygous Ref', color=colors[0])\n",
    "axes[0,1].bar(tix, ti_svHet['Heterozygous'], bottom=ti_svHet['Homozygous Ref'], label='Heterozygous', color=colors[1])\n",
    "axes[0,1].bar(tix, ti_svHet['Homozygous Alt'], bottom=ti_svHet['Homozygous Ref'] + ti_svHet['Heterozygous'], label='Homozygous Alt', color=colors[2])\n",
    "axes[0,1].set_xticks([])\n",
    "axes[0,1].set_title('B)', fontsize = 20, loc='left')\n",
    "axes[0,1].set_ylim(0,1800)\n",
    "\n",
    "axes[1,0].bar(kix, kiLC_svHet['Homozygous Ref'], label='Homozygous Ref', color=colors[0])\n",
    "axes[1,0].bar(kix, kiLC_svHet['Heterozygous'], bottom=kiLC_svHet['Homozygous Ref'], label='Heterozygous', color=colors[1])\n",
    "axes[1,0].bar(kix, kiLC_svHet['Homozygous Alt'], bottom=kiLC_svHet['Homozygous Ref'] + kiLC_svHet['Heterozygous'], label='Homozygous Alt', color=colors[2])\n",
    "axes[1,0].set_xticks([])\n",
    "axes[1,0].set_title('C)', fontsize = 20, loc='left')\n",
    "axes[1,0].set_ylim(0,1800)\n",
    "\n",
    "axes[1,1].bar(kix, kiHC_svHet['Homozygous Ref'], label='Homozygous Ref', color=colors[0])\n",
    "axes[1,1].bar(kix, kiHC_svHet['Heterozygous'], bottom=kiHC_svHet['Homozygous Ref'], label='Heterozygous', color=colors[1])\n",
    "axes[1,1].bar(kix, kiHC_svHet['Homozygous Alt'], bottom=kiHC_svHet['Homozygous Ref'] + kiHC_svHet['Heterozygous'], label='Homozygous Alt', color=colors[2])\n",
    "axes[1,1].set_xticks([])\n",
    "axes[1,1].set_title('D)', fontsize = 20, loc='left')\n",
    "axes[1,1].set_ylim(0,1800)\n",
    "\n",
    "plt.savefig('plots/Figure_2a.pdf', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kiSVmendel = kiSVmendel[kiSVmendel['Data set']!='50xSVs_10xGenotypes']\n",
    "\n",
    "order = ['10xSVs_10xGenotypes', '50xSVs_50xGenotypes']\n",
    "\n",
    "plt.figure(figsize=(10, 5))\n",
    "\n",
    "sns.violinplot(data=kiSVmendel, x='Data set', y='Percent Failing', color='0.8', order=order)\n",
    "sns.stripplot(data=kiSVmendel, x='Data set', y='Percent Failing', jitter=True, color='black', order=order)\n",
    "plt.xticks(['10xSVs_10xGenotypes', '50xSVs_50xGenotypes'], ['KĪ low', 'KĪ high'], fontsize=14)\n",
    "plt.yticks(fontsize=14)\n",
    "plt.ylim(0, 18)\n",
    "plt.ylabel('Percent Failing Sites', fontsize = 18)\n",
    "plt.xlabel('')\n",
    "plt.title('E)', fontsize = 20, loc='left')\n",
    "\n",
    "plt.savefig('plots/Figure_2b.pdf', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, we are briefly examining the average proportion of SVs in either a homozygous reference or alternate state in Australian fairy tern, tara iti, and subsampled kakī data sets.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sv_het = pd.read_csv('graphtyper/individual_svHet.tsv', delimiter='\\t')\n",
    "\n",
    "print(\"Mean proportion of homozygous alternate sites in AFT: \", sv_het[sv_het['Population']=='AU']['Proportion Alt'].mean())\n",
    "print(\"Mean proportion of homozygous alternate sites in TI: \", sv_het[sv_het['Population']=='TI']['Proportion Alt'].mean())\n",
    "print(\"Mean proportion of homozygous alternate sites in KI low: \", sv_het[sv_het['Population']=='KI_10x']['Proportion Alt'].mean())\n",
    "print(\"Mean proportion of homozygous alternate sites in KI high: \", sv_het[sv_het['Population']=='KI']['Proportion Alt'].mean())\n",
    "\n",
    "print(\"Mean proportion of homozygous reference sites in AFT: \", sv_het[sv_het['Population']=='AU']['Proportion Ref'].mean())\n",
    "print(\"Mean proportion of homozygous reference sites in TI: \", sv_het[sv_het['Population']=='TI']['Proportion Ref'].mean())\n",
    "print(\"Mean proportion of homozygous reference sites in KI low: \", sv_het[sv_het['Population']=='KI_10x']['Proportion Ref'].mean())\n",
    "print(\"Mean proportion of homozygous reference sites in KI high: \", sv_het[sv_het['Population']=='KI']['Proportion Ref'].mean())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "indiv_het = pd.read_csv('angsd/individual_het.tsv', delimiter='\\t')\n",
    "indiv_het = indiv_het.sort_values(by=['Sample', 'Population'], ascending=[True, True])\n",
    "neutral_het = indiv_het[indiv_het['Tool'] == 'neutral']\n",
    "\n",
    "neutral_het.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For relative comparisons of diversity, examined individual heterozygosity for SNPs (A) in Australian fairy tern, tara iti and high coverage kakī data sets, and SVs (B) jointly called and genotyped in fairy terns (i.e., including sites fixed in either population), as well as individual ROH counts (C) and ROH proportions (D) for all three groups.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "indiv_het = pd.read_csv('angsd/individual_het.tsv', delimiter='\\t')\n",
    "indiv_het = indiv_het.sort_values(by=['Sample', 'Population'], ascending=[True, True])\n",
    "neutral_het = indiv_het[indiv_het['Tool'] == 'neutral']\n",
    "\n",
    "sv_het = pd.read_csv('graphtyper/individual_svHet.tsv', delimiter='\\t')\n",
    "sv_fairyhet = sv_het[\n",
    "                (sv_het['Population']!='KI_10x') & \n",
    "                (sv_het['Population']!='50xSVs_10xGTs') & \n",
    "                (sv_het['Population']!='50xSVs_50xGTs') & \n",
    "                (sv_het['Population']!='AU_PlotCritic') &\n",
    "                (sv_het['Population']!='AU_noFixed') &\n",
    "                (sv_het['Population']!='AU_sites') & \n",
    "                (sv_het['Population']!='TI_PlotCritic') & \n",
    "                (sv_het['Population']!='TI_sites') &\n",
    "                (sv_het['Population']!='KI')\n",
    "                ]\n",
    "\n",
    "sv_het = sv_het[\n",
    "                (sv_het['Population']!='50xSVs_10xGTs') & \n",
    "                (sv_het['Population']!='50xSVs_50xGTs') & \n",
    "                (sv_het['Population']!='AU_PlotCritic') &\n",
    "                (sv_het['Population']!='AU') &\n",
    "                (sv_het['Population']!='AU_sites') & \n",
    "                (sv_het['Population']!='TI_PlotCritic') &\n",
    "                (sv_het['Population']!='TI')\n",
    "                ]\n",
    "\n",
    "roh = pd.read_csv('ROHAN_out/ROHs_v2.tsv', sep='\\t')\n",
    "roh = roh[(roh['Sample']!='SP01') & roh['Sample']!='TI22']\n",
    "\n",
    "roh_summary = roh[['Sample', 'Size Class', 'Population']]\n",
    "roh_summary = roh_summary.groupby(['Sample', 'Size Class', 'Population']).size().reset_index(name='ROH Counts')\n",
    "roh_summary = roh_summary.sort_values('Size Class', ascending=False)\n",
    "\n",
    "roh_len = roh[['Sample', 'ROH Length', 'Size Class', 'Population']]\n",
    "roh_len = roh_len.groupby(['Sample', 'Size Class', 'Population'])['ROH Length'].sum().reset_index(name='ROH Total')\n",
    "\n",
    "roh_conditions = [\n",
    "    roh_len['Population']=='AU',\n",
    "    roh_len['Population']=='TI',\n",
    "    roh_len['Population']=='KI_10x',\n",
    "    roh_len['Population']=='KI'\n",
    "]\n",
    "\n",
    "genome_vals = [\n",
    "    roh_len['ROH Total'] / 1088797119,\n",
    "    roh_len['ROH Total'] / 1088797119,\n",
    "    roh_len['ROH Total'] / 1095624494,\n",
    "    roh_len['ROH Total'] / 1095624494\n",
    "]\n",
    "\n",
    "roh_len['Proportion'] = np.select(roh_conditions, genome_vals, default=np.nan)\n",
    "\n",
    "roh_len = roh_len.sort_values('Size Class', ascending=False)\n",
    "\n",
    "xvalues = [\"Short ROHs\", \"Medium ROHs\", \"Long ROHs\"]\n",
    "palette = ['gold', 'steelblue', 'grey', 'black']\n",
    "order = ['AU', 'TI', 'KI_10x', 'KI']\n",
    "\n",
    "fig, axes = plt.subplots(1, 5, figsize=(42, 10), sharex=False, sharey=False)\n",
    "\n",
    "# SNP heterozygosity\n",
    "sns.violinplot(data = neutral_het, x = \"Population\", y = \"Heterozygosity\", order=order, color=\"0.8\", ax=axes[0])\n",
    "sns.stripplot(data = neutral_het, x = \"Population\", y = \"Heterozygosity\", hue = \"Population\", hue_order=['AU', 'TI', 'KI_10x', 'KI'], palette=['gold', 'steelblue', 'grey', 'black'], jitter = True, size=5, ax=axes[0])\n",
    "axes[0].set_title('A)', loc='left', fontsize=42)\n",
    "axes[0].set_xticklabels(['AFT', 'TI', 'KĪ low', 'KĪ high'], fontsize=22, rotation=45)\n",
    "axes[0].set_xlabel('Population', fontsize=22)\n",
    "axes[0].set_ylabel(\"Heterozygosity\", fontsize=20)\n",
    "axes[0].tick_params(axis='y', which='major', labelsize=18)\n",
    "\n",
    "# Fairy SV heterozygosity\n",
    "sns.violinplot(data = sv_fairyhet, x = 'Population', y = 'Heterozygosity', color='0.8', ax=axes[1])\n",
    "sns.stripplot(data = sv_fairyhet, x = 'Population', y = 'Heterozygosity', hue = 'Population', palette=palette, jitter = True, size=5, ax=axes[1])\n",
    "axes[1].set_title('B)', loc='left', fontsize=42)\n",
    "axes[1].set_xticklabels(['AFT', 'TI'], fontsize=22)\n",
    "axes[1].set_xlabel('Population', fontsize=22)\n",
    "axes[1].set_ylabel('Heterozygosity', fontsize=20)\n",
    "axes[1].tick_params(axis='y', which='major', labelsize=18)\n",
    "\n",
    "# Variable SV heterozygosity\n",
    "sns.violinplot(data = sv_het, x = 'Population', y = 'Heterozygosity', color='0.8', ax=axes[2])\n",
    "sns.stripplot(data = sv_het, x = 'Population', y = 'Heterozygosity', hue = 'Population', palette=palette, jitter = True, size=5, ax=axes[2])\n",
    "axes[2].set_title('C)', loc='left', fontsize=42)\n",
    "axes[2].set_xticklabels(['AFT', 'TI', 'KĪ low', 'KĪ high'], fontsize=20, rotation=45)\n",
    "axes[2].set_xlabel('Population', fontsize=22)\n",
    "axes[2].set_ylabel('Heterozygosity', fontsize=20)\n",
    "axes[2].tick_params(axis='y', which='major', labelsize=18)\n",
    "\n",
    "# ROH Count\n",
    "sns.boxplot(data=roh_summary, x=\"Size Class\", y=\"ROH Counts\", hue=\"Population\", palette=palette, hue_order=order, boxprops={'alpha': 0.4},ax=axes[3], legend=False)\n",
    "sns.stripplot(data=roh_summary, x=\"Size Class\", y=\"ROH Counts\", hue=\"Population\", palette=palette, hue_order=order, dodge=True, ax=axes[3], legend=False)\n",
    "axes[3].set_title('D)', loc='left', fontsize=42)\n",
    "axes[3].set_xticklabels(['Short', 'Medium', 'Large'], fontsize=20)\n",
    "axes[3].set_xlabel('ROH Size Class', fontsize=22)\n",
    "axes[3].set_ylabel(\"Count per Individual\", fontsize=20)\n",
    "axes[3].tick_params(axis='y', which='major', labelsize=18)\n",
    "\n",
    "# ROH Proportion\n",
    "sns.boxplot(data=roh_len, x=\"Size Class\", y=\"Proportion\", hue=\"Population\", palette=palette, hue_order=order, boxprops={'alpha': 0.4}, ax=axes[4], legend=False)\n",
    "sns.stripplot(data=roh_len, x=\"Size Class\", y=\"Proportion\", hue=\"Population\", palette=palette, hue_order=order, dodge=True, ax=axes[4], legend=False)\n",
    "axes[4].set_title('E)', loc='left', fontsize=42)\n",
    "axes[4].set_xticklabels(['Short', 'Medium', 'Large'], fontsize=22)\n",
    "axes[4].set_xlabel('ROH Size Class', fontsize=22)\n",
    "axes[4].set_ylabel(\"$F_{ROH}$\", fontsize=22)\n",
    "axes[4].tick_params(axis='y', which='major', labelsize=18)\n",
    "\n",
    "plt.savefig('plots/Figure_3.png', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Significance of differences in individual heterozygosity.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mann_whitney_test(df, group_col, value_col, group1, group2):\n",
    "    group1_values = df[df[group_col] == group1][value_col]\n",
    "    group2_values = df[df[group_col] == group2][value_col]\n",
    "    stat, p_value = mannwhitneyu(group1_values, group2_values, alternative='two-sided')\n",
    "    return stat, p_value\n",
    "\n",
    "groups = ['AU', 'TI', 'KI_10x', 'KI']\n",
    "results_total = {}\n",
    "\n",
    "for i in range(len(groups)):\n",
    "    for j in range(i + 1, len(groups)):\n",
    "        group1 = groups[i]\n",
    "        group2 = groups[j]\n",
    "        stat, p_value = mann_whitney_test(neutral_het, 'Population', 'Heterozygosity', group1, group2)\n",
    "        results_total[f'{group1} vs {group2}'] = {'stat': stat, 'p_value': p_value}\n",
    "\n",
    "# Print the results\n",
    "print(\"Mann-Whitney U Test results for SNP heterozygosity:\")\n",
    "for comparison, result in results_total.items():\n",
    "    print(f\"{comparison}: U-statistic = {result['stat']}, p-value = {result['p_value']}\")\n",
    "\n",
    "for i in range(len(groups)):\n",
    "    standard_dev = neutral_het[neutral_het['Population']==groups[i]]['Heterozygosity'].std()\n",
    "    count = neutral_het[neutral_het['Population']==groups[i]]['Heterozygosity'].count()\n",
    "    standard_err = standard_dev / np.sqrt(count)\n",
    "    print(f\"Mean individual SNP heterozygosity for {groups[i]}: {neutral_het[neutral_het['Population']==groups[i]]['Heterozygosity'].mean()} +/- {standard_err}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sv_het = pd.read_csv('graphtyper/individual_svHet.tsv', delimiter='\\t')\n",
    "\n",
    "def mann_whitney_test(df, group_col, value_col, group1, group2):\n",
    "    group1_values = df[df[group_col] == group1][value_col]\n",
    "    group2_values = df[df[group_col] == group2][value_col]\n",
    "    stat, p_value = mannwhitneyu(group1_values, group2_values, alternative='two-sided')\n",
    "    return stat, p_value\n",
    "\n",
    "groups = ['AU', 'TI', 'KI_10x', 'KI', 'AU_sites', 'TI_sites']\n",
    "results_total = {}\n",
    "\n",
    "for i in range(len(groups)):\n",
    "    for j in range(i + 1, len(groups)):\n",
    "        group1 = groups[i]\n",
    "        group2 = groups[j]\n",
    "        stat, p_value = mann_whitney_test(sv_het, 'Population', 'Heterozygosity', group1, group2)\n",
    "        results_total[f'{group1} vs {group2}'] = {'stat': stat, 'p_value': p_value}\n",
    "\n",
    "# Print the results\n",
    "print(\"Mann-Whitney U Test results for SV heterozygosity:\")\n",
    "for comparison, result in results_total.items():\n",
    "    print(f\"{comparison}: U-statistic = {result['stat']}, p-value = {result['p_value']}\")\n",
    "\n",
    "for i in range(len(groups)):\n",
    "    standard_dev = sv_het[sv_het['Population']==groups[i]]['Heterozygosity'].std()\n",
    "    count = sv_het[sv_het['Population']==groups[i]]['Heterozygosity'].count()\n",
    "    standard_err = standard_dev / np.sqrt(count)\n",
    "    print(f\"Mean individual SV heterozygosity for {groups[i]}: {sv_het[sv_het['Population']==groups[i]]['Heterozygosity'].mean()} +/- {standard_err}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### *F<sub>ROH</sub>* Relationship with H<sub>o</sub>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_trendline_and_r2_by_hue(ax, df, x_col, y_col, hue_col):\n",
    "    trendline_labels = []\n",
    "    for hue_value in df[hue_col].unique():\n",
    "        subset = df[df[hue_col] == hue_value]\n",
    "        # Fit the linear regression model\n",
    "        slope, intercept, r_value, p_value, std_err = linregress(subset[x_col], subset[y_col])\n",
    "        trendline = slope * subset[x_col] + intercept\n",
    "        ax.plot(subset[x_col], trendline, label=f'{hue_value} trendline', linewidth=2)\n",
    "        \n",
    "        # Collect trendline labels for a separate legend\n",
    "        trendline_labels.append(f'{hue_value} trendline')\n",
    "\n",
    "        # Calculate R^2\n",
    "        r_squared = r_value ** 2\n",
    "        \n",
    "        # Display R^2 on the plot\n",
    "        ax.text(0.05, 0.95 - 0.05 * list(df[hue_col].unique()).index(hue_value), f'{hue_value} $R^2 = {r_squared:.2f}$', transform=ax.transAxes, fontsize=14, verticalalignment='bottom')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import linregress\n",
    "\n",
    "def add_trendline_and_r2_by_hue(ax, df, x_col, y_col, hue_col):\n",
    "    # Initialize vertical placement for R^2 labels\n",
    "    vertical_offset = 0.05\n",
    "    hue_order = sorted(df[hue_col].unique())  # Ensure consistent order for hues\n",
    "    \n",
    "    for hue_index, hue_value in enumerate(hue_order):\n",
    "        subset = df[df[hue_col] == hue_value]\n",
    "        \n",
    "        # Fit the linear regression model\n",
    "        slope, intercept, r_value, p_value, std_err = linregress(subset[x_col], subset[y_col])\n",
    "        trendline = slope * subset[x_col] + intercept\n",
    "        ax.plot(subset[x_col], trendline, label=f'{hue_value} trendline', linewidth=2)\n",
    "        \n",
    "        # Calculate R^2\n",
    "        r_squared = r_value ** 2\n",
    "        \n",
    "        # Display R^2 on the plot in the bottom-left corner\n",
    "        ax.text(\n",
    "            0.05, \n",
    "            0.05 + vertical_offset * hue_index,  # Offset each label vertically\n",
    "            f'{hue_value}: $R^2 = {r_squared:.2f}$',\n",
    "            transform=ax.transAxes,\n",
    "            fontsize=10,\n",
    "            verticalalignment='bottom',\n",
    "            horizontalalignment='left'\n",
    "        )\n",
    "\n",
    "    # Add a legend for the trendlines\n",
    "    ax.legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "palette = ['gold', 'steelblue', 'grey', 'black']\n",
    "\n",
    "roh = pd.read_csv('ROHAN_out/ROHs_v2.tsv', sep='\\t')\n",
    "\n",
    "roh_len = roh.groupby(['Sample', 'Size Class', 'Population'])['ROH Length'].sum().reset_index(name='ROH Total')\n",
    "roh_len['Froh'] = roh_len['ROH Total'] / 1088797119\n",
    "\n",
    "het = pd.read_csv('angsd/individual_het.tsv', delimiter='\\t')\n",
    "het = het[het['Tool'] == 'neutral']\n",
    "\n",
    "sv_het = pd.read_csv('graphtyper/individual_svHet.tsv', delimiter='\\t')\n",
    "\n",
    "# Merge data\n",
    "diversity = pd.merge(het, roh_len, on=['Sample', 'Population'])\n",
    "svdiversity = pd.merge(sv_het, roh_len, on=['Sample', 'Population'])\n",
    "totalhet = pd.merge(het, sv_het, on=['Sample', 'Population'])\n",
    "\n",
    "# Plot\n",
    "order= ['AU', 'TI', 'KI_10x', 'KI']\n",
    "\n",
    "fig, ax = plt.subplots(1, 3, figsize=(20, 5), sharex=False, sharey=False)\n",
    "\n",
    "sns.scatterplot(diversity.groupby(['Sample', 'Population', 'Heterozygosity'])['Froh'].sum().reset_index(), x=\"Froh\", y=\"Heterozygosity\", hue=\"Population\", palette=palette, hue_order=order, s=100, edgecolor='none', ax=ax[0])\n",
    "ax[0].legend(title='Population', loc='upper right')\n",
    "ax[0].set_xlabel(\"$F_{ROH}$\", fontsize=25)\n",
    "ax[0].set_ylabel('SNP Heterozygosity', fontsize=25)\n",
    "ax[0].set_title('A)', fontsize=25, loc='left')\n",
    "add_trendline_and_r2_by_hue(ax[0], diversity.groupby(['Sample', 'Population', 'Heterozygosity'])['Froh'].sum().reset_index(), 'Froh', 'Heterozygosity', 'Population')\n",
    "\n",
    "sns.scatterplot(svdiversity.groupby(['Sample', 'Population', 'Heterozygosity'])['Froh'].sum().reset_index(), x=\"Froh\", y=\"Heterozygosity\", hue=\"Population\", palette=palette, hue_order=order, s=100, edgecolor='none', ax=ax[1])\n",
    "ax[1].legend(title='Population', loc='upper right')\n",
    "ax[1].set_xlabel(\"$F_{ROH}$\", fontsize=25)\n",
    "ax[1].set_ylabel('SV Heterozygosity', fontsize=25)\n",
    "ax[1].set_title('B)', fontsize=25, loc='left')\n",
    "add_trendline_and_r2_by_hue(ax[1], svdiversity.groupby(['Sample', 'Population', 'Heterozygosity'])['Froh'].sum().reset_index(), 'Froh', 'Heterozygosity', 'Population')\n",
    "\n",
    "sns.scatterplot(data=totalhet, x='Heterozygosity_x', y='Heterozygosity_y', hue='Population', palette=palette, hue_order=order, s=100, edgecolor='none', ax=ax[2])\n",
    "ax[2].legend(loc='upper right')\n",
    "ax[2].set_xlabel('SNP Heterozygosity', fontsize=25)\n",
    "ax[2].set_ylabel('SV Heterozygosity', fontsize=25)\n",
    "ax[2].set_title('C)', fontsize=25, loc='left')\n",
    "\n",
    "# Remove individual legends\n",
    "for a in ax:\n",
    "    a.legend_.remove()\n",
    "\n",
    "# Add a single, shared legend\n",
    "handles, labels = ax[0].get_legend_handles_labels()\n",
    "fig.legend(\n",
    "    handles, labels, title='Population', loc='upper center', bbox_to_anchor=(0.5, -0.1), \n",
    "    ncol=len(order), fontsize=14\n",
    ")\n",
    "\n",
    "plt.xticks(fontsize = 18)\n",
    "plt.yticks(fontsize = 18)\n",
    "\n",
    "# Save the plot\n",
    "plt.savefig('plots/Supp_Fig_roh_het_relationship.png', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sv_het = pd.read_csv('graphtyper/individual_svHet.tsv', delimiter='\\t')\n",
    "het = pd.read_csv('angsd/individual_het.tsv', delimiter='\\t')\n",
    "het = het[het['Tool'] == 'neutral']\n",
    "\n",
    "totalhet = pd.merge(het, sv_het, on=['Sample', 'Population'])\n",
    "\n",
    "totalhet[totalhet['Population']=='KI_10x'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Local Heterozygosity \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "for POP in AU TI KI\n",
    "    do\n",
    "    if [[ \"$POP\" == \"AU\" ]]\n",
    "        then\n",
    "        angsd -bam ${ANGSD}${POP}.list -ref $TREF -anc $TANC -out ${ANGSD}samtools/heterozygosity/${POP}_perSite \\\n",
    "            -uniqueOnly 1 -remove_bads 1 -only_proper_pairs 1 -trim 0 -C 50 -baq 1 \\\n",
    "            -minMapQ 20 -minQ 20 -minInd 19 -setMinDepth 200 -setMaxDepth 420 -doCounts 1 \\\n",
    "            -GL 1 -doHWE 1 -domajorminor 1 -doMaf 1 -SNP_pval 1e-6\n",
    "        angsd -bam ${ANGSD}${POP}.list -ref $TREF -anc $TANC -sites $TSITES -out ${ANGSD}samtools/heterozygosity/${POP}_perSite \\\n",
    "            -uniqueOnly 1 -remove_bads 1 -only_proper_pairs 1 -trim 0 -C 50 -baq 1 \\\n",
    "            -minMapQ 20 -minQ 20 -minInd 19 -setMinDepth 200 -setMaxDepth 420 -doCounts 1 \\\n",
    "            -GL 1 -doHWE 1 -domajorminor 1 -doMaf 1 -SNP_pval 1e-6\n",
    "    elif [[ \"$POP\" == \"TI\" ]]\n",
    "        then\n",
    "        angsd -bam ${ANGSD}${POP}.list -ref $TREF -anc $TANC -out ${ANGSD}samtools/heterozygosity/${POP}_perSite \\\n",
    "            -uniqueOnly 1 -remove_bads 1 -only_proper_pairs 1 -trim 0 -C 50 -baq 1 \\\n",
    "            -minMapQ 20 -minQ 20 -minInd 15 -setMinDepth 120 -setMaxDepth 280 -doCounts 1 \\\n",
    "            -GL 1 -doHWE 1 -domajorminor 1 -doMaf 1 -SNP_pval 1e-6\n",
    "        angsd -bam ${ANGSD}${POP}.list -ref $TREF -anc $TANC -sites $TSITES -out ${ANGSD}samtools/heterozygosity/${POP}_perSite \\\n",
    "            -uniqueOnly 1 -remove_bads 1 -only_proper_pairs 1 -trim 0 -C 50 -baq 1 \\\n",
    "            -minMapQ 20 -minQ 20 -minInd 15 -setMinDepth 120 -setMaxDepth 280 -doCounts 1 \\\n",
    "            -GL 1 -doHWE 1 -domajorminor 1 -doMaf 1 -SNP_pval 1e-6\n",
    "    else\n",
    "        angsd -bam ${ANGSD}${POP}.list -ref $KREF -anc $KANC -out ${ANGSD}samtools/heterozygosity/${POP}_perSite \\\n",
    "            -uniqueOnly 1 -remove_bads 1 -only_proper_pairs 1 -trim 0 -C 50 -baq 1 \\\n",
    "            -minMapQ 20 -minQ 20 -minInd 24 -setMinDepth 700 -setMaxDepth 1200 -doCounts 1 \\\n",
    "            -GL 1 -doHWE 1 -domajorminor 1 -doMaf 1 -SNP_pval 1e-6\n",
    "        angsd -bam ${ANGSD}${POP}.list -ref $KREF -anc $KANC -sites $KSITES -out ${ANGSD}neutral/heterozygosity/${POP}_perSite \\\n",
    "            -uniqueOnly 1 -remove_bads 1 -only_proper_pairs 1 -trim 0 -C 50 -baq 1 \\\n",
    "            -minMapQ 20 -minQ 20 -minInd 24 -setMinDepth 700 -setMaxDepth 1200 -doCounts 1 \\\n",
    "            -GL 1 -doHWE 1 -domajorminor 1 -doMaf 1 -SNP_pval 1e-6\n",
    "    fi\n",
    "done"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
